File: app/src/core/EventEmiter.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ArrayPushArray = ____lualib.__TS__ArrayPushArray
local __TS__Delete = ____lualib.__TS__Delete
local __TS__ArrayIndexOf = ____lualib.__TS__ArrayIndexOf
local __TS__ArraySplice = ____lualib.__TS__ArraySplice
local ____exports = {}
____exports.EventEmiter = __TS__Class()
local EventEmiter = ____exports.EventEmiter
EventEmiter.name = "EventEmiter"
function EventEmiter.prototype.____constructor(self)
    self.itemSet = {}
    self.emitItems = {}
end
function EventEmiter.prototype.beginInterceptor(self)
    local vs = {}
    local ____self_emitItems_0 = self.emitItems
    ____self_emitItems_0[#____self_emitItems_0 + 1] = vs
    return function()
        for ____, v in ipairs(vs) do
            v()
        end
    end
end
function EventEmiter.prototype.endInterceptor(self)
    table.remove(self.emitItems)
end
function EventEmiter.prototype.emit(self, name, ...)
    local args = {...}
    if #self.emitItems > 0 then
        local vs = self.emitItems[#self.emitItems]
        vs[#vs + 1] = function()
            self:_emit(
                name,
                table.unpack(args)
            )
        end
        return
    end
    self:_emit(
        name,
        table.unpack(args)
    )
end
function EventEmiter.prototype._emit(self, name, ...)
    if self.itemSet == nil or self.itemSet == nil then
        return
    end
    local handlers = {}
    local eventHandlers = self.itemSet[name]
    if eventHandlers ~= nil then
        __TS__ArrayPushArray(handlers, eventHandlers)
    end
    if name ~= "*" then
        local wildcardHandlers = self.itemSet["*"]
        if wildcardHandlers ~= nil then
            __TS__ArrayPushArray(handlers, wildcardHandlers)
        end
    end
    for ____, handler in ipairs(handlers) do
        handler(...)
    end
end
function EventEmiter.prototype.on(self, name, fn)
    if not self.itemSet[name] then
        self.itemSet[name] = {}
    end
    local ____self_itemSet_name_1 = self.itemSet[name]
    ____self_itemSet_name_1[#____self_itemSet_name_1 + 1] = fn
end
function EventEmiter.prototype.off(self, name, fn)
    if not name then
        self.itemSet = {}
        return
    end
    if not fn then
        __TS__Delete(self.itemSet, name)
        return
    end
    local handlers = self.itemSet[name]
    if handlers ~= nil then
        local index = __TS__ArrayIndexOf(handlers, fn)
        if index ~= -1 then
            __TS__ArraySplice(handlers, index, 1)
        end
        if #handlers == 0 then
            __TS__Delete(self.itemSet, name)
        end
    end
end
return ____exports

File: src/core/EventEmiter.ts

export type EventHandler = (this: void, ...args: any[]) => void
export type EventInterceptor = (this: void) => void

export interface IEventEmiter {
    emit(name: string, ...args: any[]): void
    on(name: string, fn: EventHandler): void
    off(name?: string, fn?: EventHandler): void
    beginInterceptor(): EventInterceptor
    endInterceptor(): void
}

export class EventEmiter implements IEventEmiter {
    private itemSet: { [key: string]: EventHandler[] } = {};
    private readonly emitItems: EventInterceptor[][] = []

    beginInterceptor(): EventInterceptor {
        let vs: EventInterceptor[] = []
        this.emitItems.push(vs)
        return () => {
            for (let v of vs) {
                v()
            }
        }
    }

    endInterceptor(): void {
        this.emitItems.pop()
    }

    emit(name: string, ...args: any[]): void {
        if (this.emitItems.length > 0) {
            let vs = this.emitItems[this.emitItems.length - 1]
            vs.push(() => {
                this._emit(name, ...args)
            })
            return
        }
        this._emit(name, ...args)
    }

    private _emit(name: string, ...args: any[]): void {

        if (this.itemSet === null || this.itemSet === undefined) {
            return;
        }

        const handlers: EventHandler[] = [];

        // Add handlers for the specific event
        const eventHandlers = this.itemSet[name];
        if (eventHandlers !== undefined) {
            handlers.push(...eventHandlers);
        }

        // If not a wildcard event, also add wildcard handlers
        if (name !== "*") {
            const wildcardHandlers = this.itemSet["*"];
            if (wildcardHandlers !== undefined) {
                handlers.push(...wildcardHandlers);
            }
        }

        // Execute all collected handlers
        for (const handler of handlers) {
            handler(...args);
        }
    }

    on(name: string, fn: EventHandler): void {
        if (!this.itemSet[name]) {
            this.itemSet[name] = [];
        }
        this.itemSet[name].push(fn);
    }

    off(name?: string, fn?: EventHandler): void {
        if (!name) {
            this.itemSet = {};
            return;
        }

        if (!fn) {
            delete this.itemSet[name];
            return;
        }

        const handlers = this.itemSet[name];
        if (handlers !== undefined) {
            const index = handlers.indexOf(fn);
            if (index !== -1) {
                handlers.splice(index, 1);
            }
            if (handlers.length === 0) {
                delete this.itemSet[name];
            }
        }
    }
}